### ISSUE TITLE
Improve PR reviewer efficiency through improved label automation

### ISSUE THREAD

**Issue by dtzar** [kubernetes/community#6999]
### Describe the issue

Today there is a general challenge of having far more PR contributions than there are people to review them.  Time is wasted for reviewers to determine which PRs are actually ready for review and contributors suffer with longer times to get reviews.  The process today goes as follows:

1. PR created and gets auto-assigned to reviewers list in OWNERS file. 
2. Reviewers do initial review and typically either want review from another person and/or have feedback for author to address
3. Author addresses feedback from reviewer
4. Reviewer approves
5. Approvers approve
6. All tests pass and merge happens

The wasted time occurs in this flow typically at:

1. Between 2 and 3 - the reviewers keep checking the PRs to see if the author has addressed the feedback
2. After 3 - the author may have not caught that the tests are failing or believe the tests which are failing, so reviewers have to check this out and ping the author again.  Also possible tests fail later after an automated rebase and author/reviewers aren't aware until manually checking PR.
3. After 2 - If anyone requests a second reviewer and reviewers keep coming back to check the PR to see if 2nd reviewer reviewed the PR.
4. After 4 - the reviewer may have approved the PR, but doesn't have approval permission so author is waiting for an approver to approve/merge.

All of these things result in not only wasted time for reviewers/approvers, but also delayed time for authors in getting their PRs merged.

What's the proposed solution to help?

Add 5 additional [labels](https://github.com/kubernetes/test-infra/blob/master/label_sync/labels.md) which are automatically added based on criteria to make it more quickly visible what state the PRs are in for reviewers/approvers and authors.
1. **Needs Review** - No review has occurred on the PR yet by someone in the reviewer or approver list.
Automation criteria:
    1. Add: PR creation along with assigning people from reviewers list
    1. Remove: Once the first review happens.  
    1. Add: If reviewer then assigns someone, label gets added again (signaling they want another reviewer). 
    1. Add: Assuming the review was not approved, then after the author adds another commit after this, the label gets automatically added again.
    1. Remove: Review with approval
1. **Needs Approval** - There has been one reviewer who has approved, but the PR has not be approved by approver OWNER.
Automation criteria:
    1. Add: One review with approval exists, but no approval status is present.
    1. Add: Approval status was present, but removed (i.e. rebased)
    1. Remove: Approval status is present.
1. **Wait on Author** - Reviewer has provided feedback author needs to address
Automation criteria:
    1. Added: First review has happened with a non-approved.  This could be a single _code_ comment.
    1. Added: PR is WIP or draft status.
    1. Added: Needs rebase label is added.
    1. Added: Tests fail.
    1. Removed: Author has made a single code commit after the non-approved review
    1. Removed: Manual may be needed in various cases such as random comment in the code which isn't a valid review.
1. **Abandoned** - The PR has been in "Wait on Author" state for an extended period of time. Next step here after this is implemented would be to automatically close PRs after abandoned state label has been present for another pre-determined time period (i.e. 30 days to label, another 30 days in abandoned), but this should be in a separate issue.
Automation criteria:
    1. Add: PR has been in state "Wait on Author" for 30 days.  Automatically tag the author to prompt them for a status update or close out the PR if they no longer plan to contribute.
    1. Remove: Author adds new commit to PR
1. **Tests Fail** - One or more tests are failing
Automation criteria:
    1. Add: One or more of the tests have failure
    1. Remove: Label exists and then all tests pass

**Comment by nimbinatus** [comment]
Could the labels also enable automation to remind people who need to do something on the PR (e.g., an author to address feedback) via a comment on the PR? That probably is out of scope for this proposal, but I wonder if it might be a next step after this. (P.S. - Thanks for writing all this up and thinking it through!)

**Comment by jackfrancis** [comment]
This is great! Is the "Tests Fail" label useful because the UI that folks will use doesn't have the red "x" checkbox exposed when a PR has a failing test?

**Comment by dtzar** [comment]
> This is great! Is the "Tests Fail" label useful because the UI that folks will use doesn't have the red "x" checkbox exposed when a PR has a failing test?

Yes - it's useful because you can filter out PRs where the tests are failing so you shouldn't even have to click into it to see the red x.

**Comment by dtzar** [comment]
> > Check for existence of one approval which doesn't exist in approvers list.
> 
> PRs that touch multiple directories may require approval from multiple owners. Figuring out the criteria for this one will be hard.

I revised the automation criteria - I *assume* you could just check for existence or not of approval and if so then this wouldn't matter about checking where it came from.

**Comment by dtzar** [comment]
> also, questions:
> 
> 1. What happens if a PR has one of the above statuses, and then gets a Needs-Rebase?
> 2. Could we look at different timing on expiring abandoned PRs based on the state of the PR when abandoned?

Re: Rebase - I can't think of changing automation criteria for any of those if this shows up other than I added comment on needs approval as I *think* if needs rebase happens we remove needs approval.  If there is something I'm missing, it would be great to add in a comment as such.

Re: Abandoned PRs.  This is a great idea for another label.  I added it into the proposal just now.

**Comment by jackfrancis** [comment]
I don't really love tracking "Approved", but that's also because I don't really love pre-tagging a PR w/ `/approve` and would instead prefer that we always do this "just in time" after lgtm and pasing tests to have some control over when and in which order PRs land in main branch.

(My 2¢)

**Comment by jberkus** [comment]
Yeah, we don't need a new label for abandoned.  I was just thinking that we might want to -- for example -- auto-close PRs faster if they're Waiting On Author or Test Fail.

**Comment by SergeyKanzhelev** [comment]
The flow suggested matches how we triage PRs in [sig node](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-node/triage.md) and problems we experience with the triage. If only columns would be queryable and projects had better automation in GitHub, things would be easier and sigs would be able to experiment and adjust processes.

One thing we can try though is to write some automation for the GitHub project board trying to automate your suggested policies. This automation may be done independently from the prow bot. And we will not need to introduce labels and make the process ideal from the very beginning.

For example, in the board today we manually moving PRs to waiting on author when rebase is needed. This can be relatively easy automated. Once everything is automated we can generalize this pattern in labels. LMK if this is something you want to try together.

Using boards instead of labels will also help with multi-sig triages. For example, sig node and sig windows owns some code. It is typical that both labels will be applied. And both sigs needs to triage the PR. We have this problem with bug triage labels - does sig node `/triage accept` or sig windows? Does sig node `/triage accept` bugs reported against documentation on website or SIG Docs team? Using boards make it cleaner.

In the write up I have some comments and confusions:

- PRs are created as needs-review. However, the first step of PRs triage may be valuable. Never touched PRs looked at the first time may be routed efficiently. For example, removed from sig's board, assigned to people manually, closed as trivial edits, or placed to waiting on author if description is not sufficient. It mostly works when there is a triage meeting or anything like this. I wonder if the process can be adjusted to help with this. 
- Would needs review include all PRs, except with other labels? If so, do we even need the "needs review" label?
- is the assumption that only reviewers from owners files affect labels? Would reviews from others suffice for automation? We do not have that many reviewers in SIG Node in owners files, and many people, whos lgtms would suffice makes reviews 
- Needs Review is removed after "approved". Does it mean that reviewer is placing `/approve` instead of `/lgtm`. Or this approved means lgtm'd?
- We need more ways to get out of "Waiting on Author". Reviewer should be able to reply to comment and suggest that something is not the issue. 
- The process doesn't say when PR is assigned to approver. Is it done manually as now? Or automation will handle it? Will approver be unassigned after new commit was added and PR is back to needs reviewer?

Also as mentioned before:
- there is already stale and rotten mechanism. I'm not sure why we need Abandoned label
- there is one more problem that worth looking into is filtering out things that either done by SIG (only a single file change needed approval) or larger code organization changes that are not relevant to the SIG


One more thing to improve: today `needs-rebase` applied on event from PR (comment for example). So some older PRs are not marked as `needs-rebase` before you started looking at them. Applying `needs-rebase` continuously would be a great improvement.

Thank you for initiating this!

**Comment by BenTheElder** [comment]
"tests fail" would become the third way we're reporting test results to PRs (github built in status, github comment, and now a label), which seems excessive. 

It's also going to be highly noisy in the main kubernetes repo where we struggle with flaky tests.

I would still prefer that we stick only to the github status at some point, and if we need support for querying "are tests failing" we should ask github for that (the project is in touch with github team(s)).

Also: which tests? Not all tests are required.

---


We've also done this kind of tracking in external dashboards before, but getting the state machine right is hard.

https://gubernator.k8s.io/pr (code is in kubernetes/test-infra)


In general most parts of the project only have active combined approvers+reviewers, which is also something we need to fix, the of waiting on reviewer is collapsed to waiting on an approver in much of the code currently.

**Comment by dtzar** [comment]
@SergeyKanzhelev - the proposed labels are only a means to help automate things into a board to achieve the desired outcomes.  If there are way(s) to not have to add more labels and avoid the need for code being added into prow-bot I'm all for it as long as we can achieve the desired goals.

@kerthcet - brought this related issue up in email response: https://github.com/kubernetes/community/issues/5535 and pointed to the same sig-node triage board.

There certainly is aligned and overlapping goals.  I'll send a slack DM to you both to see if we can sync up to flush this out a bit more.

**Comment by dtzar** [comment]
@BenTheElder Thanks for the response and additional context!  As long as we can automate the tests pass/fail into the board, then for sure no need for a label.  Unfortunately, no easy way I know of to query for comments or which workflow jobs pass fail in boards.  Status field visible in GitHub would work, but where is this and how is it automated?

I get that it would amazing to have a granular "which tests fail" that we could automate, but I'm happy with high-level all tests pass/fail.

**Comment by stormqueen1990** [comment]
This is a fantastic idea!

I especially like the needs review and waiting on author labels, as I've noticed I sometimes spend a huge amount of time figuring out whether PRs are waiting for review or just waiting for the author to address some comments that were left.

**Comment by chris-short** [comment]
I really like these labels. Even if added manually by triage teams they're all great ideas.

**Comment by justinsb** [comment]
Big fan of rethinking how we do this.

Can we use the assignee field to indicate "who has the ball"?  Then https://github.com/pulls/assigned would be really useful...

**Comment by SergeyKanzhelev** [comment]
> I really like these labels. Even if added manually by triage teams they're all great ideas.

You can have a board like we have for sig node and other sigs has https://github.com/kubernetes/community/blob/master/contributors/devel/sig-node/triage.md. So you will have labels in form of columns today =).

We just discussed this with @dtzar - we will try to automate some of proposals on the board we have to see if it works.

**Comment by dtzar** [comment]
Had a great sync with @SergeyKanzhelev just now.  We 100% agree on the need for PR board status columns of:

- Needs Review
- Wait(ing) on Author
- Needs Approval

Sig-node uses these status board columns today, but it is 100% manual movement between those.

If we do have a **wait on author** AND a **needs approval** status column, then by nature everything else would fall into the **Needs Review** category/column.  

**Needs Approval** column in a simple form could be automated today by existence of **lgtm** label, so although there are other more complicated scenarios here this could certainly suffice as a simple solution which works today.

**Wait on Author** is what we agreed out of the list is the highest priority to focus on adding and automating.  We talked more about the automation criteria, and I updated the description of the issue with it.

**Comment by dtzar** [comment]
In terms of implementation, I'd like to see how much we could automate using GitHub Workflows + Actions with the new Project boards.

For instance:
1. A [PR review could trigger ](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_review) moving the PR to a project board status of "Wait on Author" using [this GitHub Action](https://github.com/leonsteinhaeuser/project-beta-automations).  
2. Upon [trigger event of new label being added to PR ](https://docs.github.com/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#label) Note: not sure if you can check for specific value? move the issue to "Wait on Author" column.

There are possibly other GitHub actions in the marketplace which will help accomplish some of the desired behavior if the above "Project beta automations" action doesn't suffice.

**Comment by MadhavJivrajani** [comment]
Hey! 👋🏼 
I really like the direction of this proposal and thank you for the detailed write up! I'm trying to make sure I understand it fully with my current knowledge of label automation in the community, here are a few points that came to mind in addition to what folks have already mentioned above, please correct me if I'm missing something!

- Regarding `Tests Fail`:
  - I agree with @BenTheElder on the noisy aspect and that we should try and stick to GitHub statuses. If the intention is to provide a way for reviewers to filter PRs based on "if one or more tests failing, then might be worth skipping over for now", then the GitHub query `is:pr is:open status:pending user-review-requested:@me reviewed-by:@me` could maybe do the trick?

- Regarding `Needs Review`, `Needs Approval` and `Wait on Author`:
  - I'm having some difficulty understanding what an "approving review" is. If you mean a review submission as "Approved" (with or without an `/lgtm /approve`), then I'm not sure that would suffice considering quite a few times reviewers  comment `/lgtm /approve` without actually submitting the review with type "Approved". There is also the case of reviewers that are not from immediate OWNERs files but from a parent OWNER file.

- Is the primary purpose of this to help with project board automation or PR reviewer efficiency or both? Asking because of what the title is and what https://github.com/kubernetes/community/issues/6999#issuecomment-1343194036 says. If the aim is to help with project board automation then its probably a different-ish discussion. But if it is for reviewer efficiency I really like the idea of "Needs Reviewer" and "Needs Approver", then considering getting the state machine right with the current way things are done and implementation details can get hairy, I'd like to suggest a simpler approach that could potentially be a first step in the right direction:
  - Instead of automating the addition and removal of these 2 labels, maybe we can add the 2 labels to our repos and then use them as a custom label, [ref](https://github.com/kubernetes/test-infra/blob/a5f008845e5094ab4e5ff4f8164a2c49496969d9/prow/plugins/label/label.go#L43
). The comment would be something like to `/label needs-review`.
  - What this would allow us to do is, a reviewer can review a PR and then write "please add the label `needs-review` by commenting `/label needs-review`. 
  - Once comments are addressed and label is applied, a reviewer now has the ability to filter along 3 dimensions for their PRs should they choose to do so: 1) presence of `needs-review` 2) SIG label 3) Area Label
  - This way we aren't restricting or changing how reviews should be done in order to satisfy automation needs for these labels.
    - In case there are reviewers/approvers who already have workflows in place that work well for them, doing things this way won't disrupt anything (just in case!).
  - Ideally the `reviewed*` filters in GitHub search are quite effective.
  - Once a reviewer gives the lgtm, they can comment something like `/label needs-approval`.

- I'm not sure if there is a simple way to handle `Wait on Author`, the current add/remove criteria will cause churn and noise due to flaky tests etc (as Ben mentioned). If we do want this label for further automation purposes, might be worth applying it by hand `/label wait-on-author` by folks who triage PRs or reviewers/approvers who circle back to a PR.

**Comment by SergeyKanzhelev** [comment]
> * Is the primary purpose of this to help with project board automation or PR reviewer efficiency or both?

I think the ideal will be to try automating this as a board automation and if it works, formalize as labels

**Comment by dtzar** [comment]
@MadhavJivrajani - the end goal is to help PR reviewers be more efficient with their time used for reviews.  Reproducible automation for PR statuses via GH projects (boards) would be the target day-to-day usage.  If we can avoid having to produce new labels via other means (i.e. GH Workflows + GH Actions, etc) to do this, great.  Automation for label creation/removal is one consistent means GH makes it easy to do built-in filtering/moving for project boards as well as a way for K8s contributors to see statuses, so it still might make sense to add some of these statuses as labels once we exhaust the board automation route.

**Comment by dtzar** [comment]
> Regarding the transition from `Wait On Author` to `Needs Review`:
> Sometimes authors might push commits for WIP, either because they are tackling comments in chunks, or because they want to trigger the tests. But it's not their intent to get a review right away.
> Maybe this transition should require a manual comment from the author for `/label needs-review` to signal that they think their PR is ready again. Otherwise there would be a lot of churn.

Yes, this would be the safest way to transition.  The only downside is if the author forgets to do that step the PR would potentially get stalled for longer.   Ideas to address your common scenario while still having some form of automation:
1. Detect the new code commit, then delay the automation of adding the label
2. Automatically add the label but enable the author to manually remove the "needs review" label if they desire after that commit.

**Comment by alculquicondor** [comment]
I think we should optimize for the resource with less availability: reviewers. Authors could be reminded to add the label through a comment, just like they are reminded when they forget to add a `sig` tag or when they need to rebase.

If you disagree with my PoV, perhaps we should at least wait to add the label for `Needs Review` until all tests passed.

**Comment by dtzar** [comment]
> I think we should optimize for the resource with less availability: reviewers. Authors could be reminded to add the label through a comment, just like they are reminded when they forget to add a `sig` tag or when they need to rebase.

That makes sense.

**Comment by sftim** [comment]
If I provide feedback on a PR, can I ask Prow to re-add **Needs Review**? I'd do this to signal that I want someone else to take a look (eg a specialist from a different SIG).

**Comment by dtzar** [comment]
The new functionality they build into the PR filters would help a lot, but unfortunately it isn't available in the project boards yet.
![Screenshot 2023-07-17 093538](https://github.com/kubernetes/community/assets/3578906/ea2053cf-f58b-4796-a489-14fc2173617a)
